// electron-forge/dist/publishers/s3.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _bluebird = require('bluebird');

var _awsSdk = require('aws-sdk');

var _awsSdk2 = _interopRequireDefault(_awsSdk);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _s = require('s3');

var _s2 = _interopRequireDefault(_s);

var _oraHandler = require('../util/ora-handler');

var _oraHandler2 = _interopRequireDefault(_oraHandler);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const d = (0, _debug2.default)('electron-forge:publish:s3');

_awsSdk2.default.util.update(_awsSdk2.default.S3.prototype, {
  addExpect100Continue: function addExpect100Continue() {
    // Hack around large upload issue: https://github.com/andrewrk/node-s3-client/issues/74
  }
});

exports.default = (() => {
  var _ref = (0, _bluebird.coroutine)(function* (artifacts, packageJSON, forgeConfig, authToken, tag) {
    const s3Config = forgeConfig.s3;
    s3Config.secretAccessKey = s3Config.secretAccessKey || authToken;
    if (!(s3Config.accessKeyId && s3Config.secretAccessKey && s3Config.bucket)) {
      throw 'In order to publish to s3 you must set the "s3.accessKeyId", "process.env.ELECTRON_FORGE_S3_SECRET_ACCESS_KEY" and "s3.bucket" properties in your forge config. See the docs for more info'; // eslint-disable-line
    }

    d('creating s3 client with options:', s3Config);

    const client = _s2.default.createClient({
      s3Client: new _awsSdk2.default.S3({
        accessKeyId: s3Config.accessKeyId || process.env.AWS_ACCESS_KEY_ID,
        secretAccessKey: s3Config.secretAccessKey || process.env.AWS_SECRET_ACCESS_KEY
      })
    });
    client.s3.addExpect100Continue = function () {};

    const folder = s3Config.folder || tag || packageJSON.version;

    let uploaded = 0;
    yield (0, _oraHandler2.default)(`Uploading Artifacts ${uploaded}/${artifacts.length}`, (() => {
      var _ref2 = (0, _bluebird.coroutine)(function* (uploadSpinner) {
        const updateSpinner = function updateSpinner() {
          uploadSpinner.text = `Uploading Artifacts ${uploaded}/${artifacts.length}`; // eslint-disable-line
        };

        yield _promise2.default.all(artifacts.map(function (artifactPath) {
          return new _promise2.default((() => {
            var _ref3 = (0, _bluebird.coroutine)(function* (resolve, reject) {
              const done = function done(err) {
                if (err) return reject(err);
                uploaded += 1;
                updateSpinner();
                resolve();
              };

              const uploader = client.uploadFile({
                localFile: artifactPath,
                s3Params: {
                  Bucket: s3Config.bucket,
                  Key: `${folder}/${_path2.default.basename(artifactPath)}`,
                  ACL: s3Config.public ? 'public-read' : 'private'
                }
              });
              d('uploading:', artifactPath);

              uploader.on('error', function (err) {
                return done(err);
              });
              uploader.on('progress', function () {
                d(`Upload Progress (${_path2.default.basename(artifactPath)}) ${Math.round(uploader.progressAmount / uploader.progressTotal * 100)}%`);
              });
              uploader.on('end', function () {
                return done();
              });
            });

            return function (_x7, _x8) {
              return _ref3.apply(this, arguments);
            };
          })());
        }));
      });

      return function (_x6) {
        return _ref2.apply(this, arguments);
      };
    })());
  });

  return function (_x, _x2, _x3, _x4, _x5) {
    return _ref.apply(this, arguments);
  };
})();
